File: implementations/java/compression/src/utils/fractal/mapping/Codebook.java
package implementations.java.compression.src.utils.fractal.mapping;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

import implementations.java.compression.src.utils.files.readers.SequenceInput;
import implementations.java.compression.src.utils.files.readers.SequenceReader;
import implementations.java.compression.src.utils.files.writers.SequenceOutput;
import implementations.java.compression.src.utils.files.writers.SequenceWriter;
import implementations.java.compression.src.utils.fractal.transformation.TransformationType;

public class Codebook {

    /** Subfolder under each stem’s {@code processes/java/<stem>/} where {@code codebook_r{r}_d{d}.fc} files live. */
    public static final String CODEBOOKS_SUBDIR = "codebooks";

    private int rangeSize;
    private int domainSize;
    private List<FractalMapping> mappings;

    private String MAGIC_NUMBER = "FC";

    public Codebook(int rangeSize, int domainSize, List<FractalMapping> mappings) {
        this.rangeSize = rangeSize;
        this.domainSize = domainSize;
        this.mappings = mappings;
    }

    /**
     * This constructor initializes the Codebook based on a path
     * on where to look for the codebook serialized file.
     * 
     * @param path Path to codebook file
     */

    public Codebook(String path) {
        this.mappings = List.of();
        try (SequenceInput sr = new SequenceReader(path)) {
            deserialize(sr);
        } catch (Exception e) {
            System.out.println("Error deserializing codebook: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }

    public List<FractalMapping> getMappings() {
        return this.mappings;
    }

    /** Range block edge length stored in the file header. */
    public int rangeSize() {
        return rangeSize;
    }

    /** Domain block edge length stored in the file header. */
    public int domainSize() {
        return domainSize;
    }

    /**
     * Canonical on-disk path for this range/domain geometry under a stem’s {@code processes/java/<stem>/} folder
     * ({@link #CODEBOOKS_SUBDIR} subdirectory).
     */
    public static Path pathForGeometry(Path runDir, int rangeSize, int domainSize) {
        return runDir.resolve(CODEBOOKS_SUBDIR).resolve(String.format("codebook_r%d_d%d.fc", rangeSize, domainSize));
    }

    private void deserialize(SequenceInput sr) throws IOException {

        // Read first two bytes. Should be magic number
        String magicNumber = sr.readString(2);

        if (!MAGIC_NUMBER.equals(magicNumber)) {
            throw new IllegalArgumentException(
                    "Magic number of image is not correct. Expected " + MAGIC_NUMBER + "but was: " + magicNumber);
        }

        sr.space();
        this.rangeSize = sr.readInt();
        this.domainSize = sr.readInt();
        int totalRows = sr.readInt();

        List<FractalMapping> mappings = new ArrayList<>();

        for (int i = 0; i < totalRows; i++) {
            FractalMapping fm = readMapping(sr);
            mappings.add(fm);
        }

        this.mappings = mappings;
    }

    /**
     * The fractal mapping is composed of 4 ints and 2 floats.
     * 
     * @param sr
     * @return
     * @throws IOException
     */
    private FractalMapping readMapping(SequenceInput sr) throws IOException {
        int rangeX = sr.readInt();
        int rangeY = sr.readInt();
        int domainX = sr.readInt();
        int domainY = sr.readInt();
        float s = sr.readFloat();
        float o = sr.readFloat();

        TransformationType t = parseTransformation(sr);

        return new FractalMapping(rangeX, rangeY, domainX, domainY, s, o, t);
    }

    private TransformationType parseTransformation(SequenceInput sr) throws IOException {
        int ordinal = sr.readInt();
        return TransformationType.values()[ordinal];
    }

    /**
     * Serializes the Codebook into a file.
     * 
     * @param path where to save the serialization
     * @throws Exception
     */
    public void save(String path) throws Exception {
        Path p = Path.of(path);
        Path parent = p.getParent();
        if (parent != null) {
            Files.createDirectories(parent);
        }

        try (SequenceOutput sw = new SequenceWriter(path)) {
            serialize(sw);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

    }

    private void serialize(SequenceOutput sw) throws IOException {

        serializeMetadata(sw);

        for (int i = 0; i < this.mappings.size(); i++) {
            FractalMapping fm = mappings.get(i);

            sw.write(fm.rangeX());
            sw.space();
            sw.write(fm.rangeY());
            sw.space();
            sw.write(fm.domainX());
            sw.space();
            sw.write(fm.domainY());
            sw.space();
            sw.write(fm.s(), 2);
            sw.space();
            sw.write(fm.o(), 2);
            sw.space();
            sw.write(fm.t().ordinal());

            sw.bl();
        }

    }

    private void serializeMetadata(SequenceOutput sw) throws IOException {
        // Write the MAGIC_NUMBER
        sw.write(MAGIC_NUMBER);

        // Write a space
        sw.space();

        sw.write(rangeSize);

        sw.space();

        sw.write(domainSize);

        sw.space();

        // total rows
        sw.write(this.mappings.size());

        // Breakline
        sw.bl();
    }

}

File: implementations/cpp/fractal/mapping/codebook.cpp
#include "codebook.hpp"

#include <cstdlib>
#include <fstream>
#include <stdexcept>
#include <string>
#include <vector>

#include "../../files/readers/sequence-reader.hpp"
#include "../../utils/run_logging.hpp"
#include "fractal-mapping.hpp"

#include "../../files/writers/sequence-writer.hpp"
#include "../../files/writers/sequence-output.hpp"

using namespace std;

namespace {

constexpr int kTransformationTypeCount = 8;

}  // namespace

Codebook::Codebook() : rangeSize(0), domainSize(0), mappings() {}

Codebook::Codebook(int rangeSize, int domainSize, vector<FractalMapping>* mappings)
    : rangeSize(rangeSize), domainSize(domainSize), mappings(*mappings) {}


void Codebook::save(string path) {
    SequenceWriter writer(path);
    if (!writer.ok()) {
        run_logging::error("could not open for write: " + path);
        std::exit(1);
    }
    serialize(writer);
}

void Codebook::read(string path) {
    ifstream fin(path);
    if (!fin) {
        throw runtime_error("could not open codebook for read: " + path);
    }
    SequenceReader sr(fin);
    deserialize(sr);
}

void Codebook::deserialize(SequenceReader& in) {
    vector<char> mag = in.read(2);
    string magic(mag.begin(), mag.end());
    if (magic != "FC") {
        throw runtime_error("codebook magic must be FC, got: " + magic);
    }
    rangeSize = in.readNextInt();
    domainSize = in.readNextInt();
    const int n = in.readNextInt();
    if (n < 0) {
        throw runtime_error("codebook mapping count must be non-negative");
    }
    vector<FractalMapping> loaded;
    loaded.reserve(static_cast<size_t>(n));
    for (int i = 0; i < n; ++i) {
        const int rx = in.readNextInt();
        const int ry = in.readNextInt();
        const int dx = in.readNextInt();
        const int dy = in.readNextInt();
        const float s = in.readNextFloat();
        const float o = in.readNextFloat();
        const int tord = in.readNextInt();
        if (tord < 0 || tord >= kTransformationTypeCount) {
            throw runtime_error("invalid transformation ordinal in codebook");
        }
        const auto t = static_cast<TransformationType>(tord);
        loaded.emplace_back(rx, ry, dx, dy, s, o, t);
    }
    mappings = std::move(loaded);
}

void Codebook::serialize(SequenceOutput& so){

    so.write("FC");
    so.space();
    so.write(rangeSize);
    so.space();
    so.write(domainSize);
    so.space();
    so.write((int)mappings.size());
    so.bl();

    for (FractalMapping mapping : mappings) {
        so.write(mapping.range_x);
        so.space();
        so.write(mapping.range_y);
        so.space();
        so.write(mapping.domain_x);
        so.space();
        so.write(mapping.domain_y);
        so.space();
        so.write(mapping.s);
        so.space();
        so.write(mapping.o);
        so.space();
        so.write(static_cast<int>(mapping.transformation_type));
        so.bl();
    }
    

}
